# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Youssef1143/flyrank/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

In [7]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
print("""
I'm going freestyle, with a Growth/Recovery/Momentum Prediction direction: instead of scoring pages by their current state,
I want to predict whether a page will decline or grow in the near future, using only what was known before that happens.
I chose this over the four predefined lanes because it's genuine forecasting rather than describing the present — it forces a proper future-window label,
a leakage audit, and time-aware/client-grouped validation, which are core,
transferable ML skills. The trade-off is that I have to build my own label from raw daily data instead of using one that's already provided, which carries more risk of subtle errors,
but that risk is also the point: real-world labels are rarely handed to you clean. """)


I'm going freestyle, with a Growth/Recovery/Momentum Prediction direction: instead of scoring pages by their current state,
I want to predict whether a page will decline or grow in the near future, using only what was known before that happens.
I chose this over the four predefined lanes because it's genuine forecasting rather than describing the present — it forces a proper future-window label,
a leakage audit, and time-aware/client-grouped validation, which are core, 
transferable ML skills. The trade-off is that I have to build my own label from raw daily data instead of using one that's already provided, which carries more risk of subtle errors,
but that risk is also the point: real-world labels are rarely handed to you clean. 


## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

In [2]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
print("""
Question:
 Given a page's search/content signals over a prior 90-day window, can we predict whether it will decline or grow over the following 30 days?

Unit of analysis:
 one content page for one client, evaluated at a specific cutoff date (so the same page can appear multiple times if I use multiple cutoff dates).

Decision this improves:
 which pages a reviewer should proactively check before they visibly decline, rather than only reacting after impressions/clicks have already dropped.

Action a human takes:
 flagged pages go into a review queue for refresh, expansion, or monitoring — the same kind of action as Lane 2, but triggered earlier, based on predicted risk rather than an already-visible drop.

Cost of a false positive:
a reviewer spends time checking a page that was never actually going to decline — wasted but low-stakes time.

Cost of a false negative:
 a page that was quietly about to decline gets skipped, and by the time it's obviously declining (Lane 2's world), some of that traffic/opportunity is already lost — this is the more expensive error, since early warning is the entire value proposition of this direction.
""")



Question:
 Given a page's search/content signals over a prior 90-day window, can we predict whether it will decline or grow over the following 30 days?

Unit of analysis:
 one content page for one client, evaluated at a specific cutoff date (so the same page can appear multiple times if I use multiple cutoff dates).

Decision this improves:
 which pages a reviewer should proactively check before they visibly decline, rather than only reacting after impressions/clicks have already dropped.

Action a human takes:
 flagged pages go into a review queue for refresh, expansion, or monitoring — the same kind of action as Lane 2, but triggered earlier, based on predicted risk rather than an already-visible drop.

Cost of a false positive: 
a reviewer spends time checking a page that was never actually going to decline — wasted but low-stakes time.

Cost of a false negative:
 a page that was quietly about to decline gets skipped, and by the time it's obviously declining (Lane 2's world), some 

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [20]:
import pandas as pd

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

decline_share = (df["trend_direction"] == "down").mean()
eligible_share = ((df["impressions_90d"] > 0) & (df["content_age_days"] >= 90)).mean()

print("Rows:", len(df))
print("Share currently declining:", round(decline_share, 3))
print("Share eligible for a 90-day window:", round(eligible_share, 3))

Rows: 30000
Share currently declining: 0.542
Share eligible for a 90-day window: 1.0


To check whether this direction is worth pursuing, I loaded the starter dataset and looked at two things: how unstable the content already is, and whether pages have enough history to support a 90-day feature window.

30,000 rows in the starter dataset — a reasonable size to sanity-check ideas before moving to the full warehouse.
54.2% of pages are currently flagged as declining (trend_direction == "down"). This means instability is common, not rare — a real, sizeable share of content is worth prioritizing, which justifies building a system to catch this earlier.
100% of pages are eligible for a 90-day window (they meet the impressions_90d > 0 and content_age_days >= 90 filters). This confirms the starter data has enough history depth to support the prior-90-day / next-30-day window structure my direction depends on.

Together these numbers show the underlying instability is common enough to matter, and there's enough historical depth in the data to build the future-window labels my approach needs — supporting that this is worth the next 7 weeks, rather than a rare or thin problem.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

In [8]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
print("""
I can claim this is an observational, decision-support prioritization tool:
it ranks pages by predicted future risk/opportunity so a human can review the most promising candidates first.
I cannot claim that any predicted decline or recovery is guaranteed,
that a refresh action would cause a recovery (that requires a causal experiment, not this data),
or that any result reflects an actual Google ranking factor. All labels are self-defined future-window outcomes built from observable signals only — never from FlyRank's internal product scores or decision flags.""")


I can claim this is an observational, decision-support prioritization tool:
it ranks pages by predicted future risk/opportunity so a human can review the most promising candidates first. 
I cannot claim that any predicted decline or recovery is guaranteed,
that a refresh action would cause a recovery (that requires a causal experiment, not this data),
or that any result reflects an actual Google ranking factor. All labels are self-defined future-window outcomes built from observable signals only — never from FlyRank's internal product scores or decision flags.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.